clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -164,22 +167,25 @@ public synchronized AuthenticationException getAndClearAuthenticationException()
      * Wait for metadata update until the current version is larger than the last version we know of
      */
     public synchronized void awaitUpdate(final int lastVersion, final long maxWaitMs) throws InterruptedException {
-        if (maxWaitMs < 0) {
+        if (maxWaitMs < 0)
             throw new IllegalArgumentException("Max time to wait for metadata updates should not be < 0 milliseconds");
-        }
+
         long begin = System.currentTimeMillis();
         long remainingWaitMs = maxWaitMs;
-        while (this.version <= lastVersion) {
+        while ((this.version <= lastVersion) && !isClosed()) {
             AuthenticationException ex = getAndClearAuthenticationException();
             if (ex != null)
                 throw ex;
             if (remainingWaitMs != 0)
-                wait(remainingWaitMs);
+                wait(Math.min(remainingWaitMs, 100));
```

Comment:
Since we have the notify in `close()`, do we still need this change?

Reply:
Good point, reverted.